omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+      if (words.front()!="0" && words.front()!="none")
+      {
+        // The user has specified attributes other than default, these will either be added to the list or replace existing default values
+        extrasString = words.front();
+        for (auto &attrib : defaultTagAttribs)
+        {
+          auto pos = words.front().find(attrib.first);
+          std::string size;
+          if (pos != std::string::npos) // if attrib is one of our defined defaults
+          {
+            std::string remaining = words.front().substr(pos + attrib.first.size(), words.front().size());
+            auto nextPos = remaining.find('&');
+            if (nextPos != std::string::npos)
+            {
+              size = remaining.substr(0, nextPos);
+              extrasString = words.front().erase(pos-1, attrib.first.size() + nextPos + 1);
+            }
+            else
+            {
+              size = remaining.substr(0, remaining.size());
+              extrasString = words.front().erase(pos-1, words.front().size()-(pos-1));
+            }
+
+            if (size == attrib.second) // if defined value is identical to default, continue
+            {
+              continue;
+            }
+            else // set new default value
+            {
+              attrib.second = size;
+            }
+          }
+        }
+
+        // we check to see if str_capacity is set, this is needed to get strings from UDTs
+        size_t pos = words.front().find("str_max_capacity=");
+        std::string size;
+        if (pos != std::string::npos)
+        {
+          std::string remaining = words.front().substr(pos + 17, words.front().size());
+          auto nextPos = remaining.find('&');
+          if (nextPos != std::string::npos)
+          {
+            size = remaining.substr(0, nextPos);
+          }
+          else
+          {
+            size = remaining.substr(0, remaining.size());
+          }
+          keyWords.at("strCapacity") = size; 
+        }
+      }
+
+      for (auto attrib : defaultTagAttribs)
+      {
+        if (attrib.first.substr(0,3) == "str" && keyWords.at("dataType") != "STRING")
+        {
+          // If the user requests str type attributes for a non STRING record then ignore
+          continue;
+        }
+        extrasString += "&";
+        extrasString += attrib.first;
+        extrasString += attrib.second;
+      }
+      keyWords.at("tagExtras") = extrasString;
+      words.pop_back();
+    }
+  }
+
+  asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Creating libplctag tag with the following parameters:\n", driverName, functionName);
+  for (auto i = keyWords.begin(); i != keyWords.end(); i++)
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s \t\t\t %s\n", i->first.c_str(), i->second.c_str());
+  }
+  return keyWords;
+}
+
+asynStatus drvOmronEIP::loadStructFile(const char * portName, const char * filePath)
+{
+  const char * functionName = "loadStructFile";
+  this->lock();
```

Comment:
Do we need to lock the asynport in this function?

Reply:
Im pretty sure this is no longer needed, so have removed it. I needed to ensure that this was run before the database files are loaded and not at the same time. If they do run at the same time then there can be a seg fault, but I dont think the lock is required for this. This should happen naturally. I have also added a check to make sure that this function can only be run before the ioc has started running.